conans/model/conan_file.py

```diff
@@ -10,44 +10,205 @@
     exports_sources = None
+
+    #: List or tuple of strings with names of generators.
     generators = []
     revision_mode = "hash"
 
-
     # Settings and Options
+
+    #: List of strings with the first level settings (from ``settings.yml``) that the recipe
+    #: need, because:
+    #:
+    #:  - They are read for building (e.g: `if self.settings.compiler == "gcc"`)
+    #:  - They affect the ``package_id``. If a value of the declared setting changes, the
+    #:    ``package_id`` has to be different.
     settings = None
+
+    #: Dictionary with traits that affects only the current recipe, where the key is the
+    #: option name and the value is a list of different values that the option can take.
+    #: By default any value change in an option, changes the ``package_id``. Check the
+    #: ``default_options`` field to define default values for the options.
     options = None
+
+    #: The attribute ``default_options`` defines the default values for the options, both for the
+    #: current recipe and for any requirement.
+    #: This attribute should be defined as a python dictionary.
     default_options = None
 
+    #: This attribute declares that the recipe provides the same functionality as other recipe(s).
+    #: The attribute is usually needed if two or more libraries implement the same API to prevent
+    #: link-time and run-time conflicts (ODR violations). One typical situation is forked libraries.
+    #: Some examples are:
+    #:
+    #: - `LibreSSL <https://www.libressl.org/>`__, `BoringSSL <https://boringssl.googlesource.com/boringssl/>`__ and `OpenSSL <https://www.openssl.org/>`__
+    #: - `libav <https://en.wikipedia.org/wiki/Libav>`__ and `ffmpeg <https://ffmpeg.org/>`__
+    #: - `MariaDB client <https://downloads.mariadb.org/client-native>`__ and `MySQL client <https://dev.mysql.com/downloads/c-api/>`__
     provides = None
+
+    #: This attribute declares that the recipe is deprecated, causing a user-friendly warning
+    #: message to be emitted whenever it is used
     deprecated = None
 
+    #: Optional.
+    #: Declaring the ``package_type`` will help Conan:
+    #:
+    #:  - To choose better the default ``package_id_mode`` for each dependency, that is, how a change
+    #:    in a dependency should affect the ``package_id`` to the current package.
+    #:  - Which information from the dependencies should be propagated to the consumers, like
+    #:    headers, libraries, runtime information...
+    #:
+    #: The valid values are:
+    #:
+    #:     - **application**: The package is an application.
+    #:     - **library**: The package is a generic library. It will try to determine
+    #:       the type of library (from ``shared-library``, ``static-library``, ``header-library``)
+    #:       reading the ``self.options.shared`` (if declared) and the ``self.options.header_only``:
+    #:
+    #:          - **shared-library**: The package is a shared library.
```

Comment:
Not sure if this needs to be inside ``library`` type, might be confusing, can be at the same level